exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/FunctionImplementationRegistry.java

```diff
@@ -186,4 +226,105 @@ public boolean isFunctionComplexOutput(String name) {
     return false;
   }
 
+  public RemoteFunctionRegistry getRemoteFunctionRegistry() {
+    return remoteFunctionRegistry;
+  }
+
+  public List<Func> validate(Path path) throws IOException {
+    URL url = path.toUri().toURL();
+    URL[] urls = {url};
+    ClassLoader classLoader = new URLClassLoader(urls);
+    return drillFuncRegistry.validate(path.getName(), scan(classLoader, path, urls));
+  }
+
+  public void register(String jarName, ScanResult classpathScan, ClassLoader classLoader) {
+    drillFuncRegistry.register(jarName, classpathScan, classLoader);
+  }
+
+  public void unregister(String jarName) {
+    drillFuncRegistry.unregister(jarName);
+  }
+
+  /**
+   * Loads all missing functions from remote registry.
+   * Compares list of already registered jars and remote jars, loads missing jars.
+   * Missing jars are stores in local DRILL_UDF_DIR.
+   *
+   * @return true if at least functions from one jar were loaded
+   */
+  public boolean loadRemoteFunctions() {
+    List<String> missingJars = Lists.newArrayList();
+    Registry registry = remoteFunctionRegistry.getRegistry();
+
+    List<String> localJars = drillFuncRegistry.getAllJarNames();
+    for (Jar jar : registry.getJarList()) {
+      if (!localJars.contains(jar.getName())) {
+        missingJars.add(jar.getName());
+      }
+    }
+
+    for (String jarName : missingJars) {
+      try {
+        Path localUdfArea = new Path(new File(getUdfDir()).toURI());
```

Comment:
I would suggest we create the local UDF dir in Java code when needed. We cannot guarantee that Drill will always be run via the scripts. (For example, it is seldom run via scripts when debugging.) Reserve the scripts for stuff that the Java code really cannot do, such as setting the environment, building the class path, etc.

The drill-config.sh approach is not wrong, it is just one more detail that must be made right to debug drill, raising the hassle-factor by a bit more. Hence the desire to do the mkdirs in Java.